yapapi/runner/__init__.py

```diff
@@ -349,56 +342,63 @@ async def start_worker(agreement: rest.market.Agreement):
+                            current_worker_task = consumer.last_item
+                            if current_worker_task:
+                                emit(
+                                    events.TaskStarted(
+                                        agr_id=agreement.id,
+                                        task_id=current_worker_task.id,
+                                        task_data=current_worker_task.data,
+                                    )
+                                )
+                            task_id = current_worker_task.id if current_worker_task else None
+                            await batch.prepare()
+                            cc = CommandContainer()
+                            batch.register(cc)
+                            remote = await act.send(cc.commands())
                             emit(
-                                events.CommandExecuted(
-                                    agr_id=agreement.id,
-                                    task_id=task.id,
-                                    message=step.message,
-                                    cmd_idx=step.idx,
+                                events.ScriptSent(
+                                    agr_id=agreement.id, task_id=task_id, cmds=cc.commands()
                                 )
                             )
-                        emit(events.GettingResults(agr_id=agreement.id, task_id=task.id))
-                        await batch.post()
-                        emit(events.ScriptFinished(agr_id=agreement.id, task_id=task.id))
-                        await accept_payment_for_agreement(agreement.id, partial=True)
-                    except Exception as exc:
-                        task.reject_task(reason=f"failure: {exc}")
-                        raise
+                            async for step in remote:
+                                emit(
+                                    events.CommandExecuted(
+                                        agr_id=agreement.id,
+                                        task_id=task_id,
+                                        message=step.message,
+                                        cmd_idx=step.idx,
+                                    )
+                                )
+                            emit(events.GettingResults(agr_id=agreement.id, task_id=task_id))
+                            await batch.post()
+                            emit(events.ScriptFinished(agr_id=agreement.id, task_id=task_id))
+                            await accept_payment_for_agreement(agreement.id, partial=True)
+                        except Exception as exc:
+                            try:
+                                await command_generator.athrow(Exception, exc)
+                            except Exception:
+                                traceback.print_exc()
+                                emit(events.WorkerFinished(agr_id=agreement.id))
+                                return
```

Comment:
I like the idea of throwing `exc` in `worker` with `athrow`.

I don't like the idea of printing the traceback in `except Exception`, it does not give much information. For example I got a number of errors when executing commands remotely and in each case the traceback is rather useless:
```
Traceback (most recent call last):
  File "/home/azawlocki/golem/yagna-integration/yapapi/yapapi/runner/__init__.py", line 398, in start_worker
    await command_generator.athrow(Exception, exc)
  File "examples/blender/blender.py", line 51, in worker
    yield ctx.commit()
TypeError: list indices must be integers or slices, not str
```
```
Traceback (most recent call last):
  File "/home/azawlocki/golem/yagna-integration/yapapi/yapapi/runner/__init__.py", line 398, in start_worker
    await command_generator.athrow(Exception, exc)
  File "examples/blender/blender.py", line 51, in worker
    yield ctx.commit()
AssertionError: invalid gftp publish response
```
```
Traceback (most recent call last):
  File "/home/azawlocki/golem/yagna-integration/yapapi/yapapi/runner/__init__.py", line 398, in start_worker
    await command_generator.athrow(Exception, exc)
  File "examples/blender/blender.py", line 51, in worker
    yield ctx.commit()
yapapi.rest.activity.CommandExecutionError: ('stderr: Local service error: Transfer error: IO error: No such file or directory (os error 2)', 5)
```
In case of errors in remote command execution I think it would be better just to emit a new `WorkerFailed` event with the error message (the command's output) (see https://github.com/golemfactory/yapapi/pull/48/files).

Reply:
Of course for exceptions raised in the `worker` function provided by user, having the traceback is very useful. Maybe we should distinguish exceptions from remote commands that are re-raised in the generator by `command_generator.athrow` and not handled there, from the exceptions originally raised in the generator? And display the traceback only for exceptions of the second kind?

Reply:
stack traces helps me. bug is not always on `commit()`. i was thinking to make tracetraceback's only if `YA_TRACEBACK` enviroment variable is set.

Reply:
Yes, stack traces are useful here in general, except for the ones from `commit()`, as in the three examples above. Maybe we should suppress them for exceptions from `commit()`?

Reply:
Clarifying further, I think it would be good to distinguish here between
1) errors in `worker` (user code)
2) errors in `YAPAPI` (our code)
3) errors in providers (due to errors in exe-unit, the daemon, random node failures)
Events from 1) and 2) should be given `ERROR`/`CRITICAL` level (in the sense of `logging`), print the stack trace etc., events from 3) are IMO `INFO/WARNING`, so no stack trace for them.